aws-s3-transfer-manager/src/operation/upload_objects.rs

```diff
@@ -24,14 +32,56 @@ pub(crate) struct UploadObjects;
 impl UploadObjects {
     /// Execute a single `UploadObjects` transfer operation
     pub(crate) async fn orchestrate(
-        _handle: Arc<crate::client::Handle>,
-        _input: UploadObjectsInput,
-    ) -> Result<UploadObjectsHandle, UploadObjectsError> {
-        unimplemented!()
+        handle: Arc<crate::client::Handle>,
+        input: UploadObjectsInput,
+    ) -> Result<UploadObjectsHandle, crate::error::Error> {
+        //  validate existence of source and return error if it's not a directory
+        let source = input.source().expect("source set");
+        validate_target_is_dir(source).await?;
+
+        let concurrency = handle.num_workers();
+        let ctx = UploadObjectsContext::new(handle.clone(), input);
+
+        // spawn all work into the same JoinSet such that when the set is dropped all tasks are cancelled.
+        let mut tasks = JoinSet::new();
+        let (work_tx, work_rx) = async_channel::bounded(concurrency);
+
+        // spawn worker to discover/distribute work
+        tasks.spawn(worker::list_directory_contents(
+            ctx.state.input.clone(),
+            work_tx,
+        ));
+
+        for i in 0..concurrency {
+            let worker = worker::upload_objects(ctx.clone(), work_rx.clone())
+                .instrument(tracing::debug_span!("object-uploader", worker = i));
+            tasks.spawn(worker);
+        }
+
+        let handle = UploadObjectsHandle { tasks, ctx };
+        Ok(handle)
     }
 }
 
-/// Error type for `UploadObjects` operation
-#[non_exhaustive]
+/// DownloadObjects operation specific state
 #[derive(Debug)]
-pub enum UploadObjectsError {}
+pub(crate) struct UploadObjectsState {
+    input: UploadObjectsInput,
```

Comment:
trivial/debatable: I feel like `input` doesn't belong in the `State` variable. This is just read-only information that should be in the context. If we remove `input` from here, we can also just wrap the state in a mutex, since nothing in the state should be touched without holding the lock, and replace Atomics with a simple `u64`, though I'm not sure if that will be better.

Reply:
Added TODO in fcfac43. If `input` is to move to `UploadObjectsContext`, then given its definition
```
type UploadObjectsContext = TransferContext<UploadObjectsState>;
```
`TransferContext` needs to have another generic parameter `I` to be able to contain either download input type or upload input type.